server: add a new `start` flag `--pid-file` that does the right thing.

knz opened the pull request:
This patch adds support for `--pid-file`: if specified with `start`,
the file is written (and possibly created) upon successful node
startup and its contents are set to the PID of the running process.

This is meant to play well with process managers that want the real
PID of the server process (as opposed to the PID of the first process
launched by the shell, which will be different from the real PID when
`--background` is also specified).

It also makes it possible to reliably wait for server startup in
scripts etc. by specifying a named FIFO as parameter to this
flag. Thanks to this opportunity the CLI interactive tests can now
avoid a sleep upon each server start.

Fixes  #13992.
Fixes #13987.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13996)
<!-- Reviewable:end -->

tamird commented:
🐶  LGTM

dt commented on pkg/cli/interactive_tests/test_missing_log_output.tcl:
```diff
@@ -63,6 +63,10 @@ eexpect ":/# "
+eexpect "\[1\]"
+eexpect ":/# "
+# Give the server some time to start.
+sleep 0.5
```
(sorry, i don't have any better ideas, but sleeping in a test makes me sad.)

knz replied:
We ought to have a mechanism that makes the server detach from the terminal, or indicates via a pipe/file its PID, only after it has started successfully. Unfortunately we don't have that.

dt replied:
we do -- `start --background` waits until the server is ready, using systemd's readiness signaling.

concretely, sleeping a fixed duration this seems likely to flake when CIs / docker scheduler is unusually slow and we happen to cross the 0.5s, or 1s, etc mark, but if we make it so long we never hit it, we have a slow test.

dt replied:
Anyway, this fix is still better than no fix, and I certainly didn't intend to let the perfect become the enemy of the good.

knz replied:
Just to think with you, is there a way to use the systemd readiness thing without systemd?

dt replied:
we call `sdnotify.Ready()` in server, so if there's a `NOTIFY_SOCKET` defined in the env, we'll write `READY=1` to it when we're running.

I'm pretty sure there's an off-the-shelf util that can run a cmd and block until that msg is received?

alternately, we can just use our own `start --background` which does that for us.

knz replied:
yes this off-the-shelf utility is called `cat`. I'll try that.

knz committed:
server: add a new `start` flag `--pid-file` that does the right thing.

This patch adds support for `--pid-file`: if specified with `start`,
the file is written (and possibly created) upon successful node
startup and its contents are set to the PID of the running process.

This is meant to play well with process managers that want the real
PID of the server process (as opposed to the PID of the first process
launched by the shell, which will be different from the real PID when
`--background` is also specified).

It also makes it possible to reliably wait for server startup in
scripts etc. by specifying a named FIFO as parameter to this
flag. Thanks to this opportunity the CLI interactive tests can now
avoid a sleep upon each server start.

knz commented:
I went for a completely different approach, which was also on my wishlist for a long time. PTAL.

tamird commented:
Nice!

knz commented:
TFYR!
